### PR TITLE
Limit fields

### PR DESCRIPTION
**Pull request by liakify**
Addresses #131 and #132 
**Changelog:**
- `Phone`s are limited to 20 digits (i.e. 3 to 20 digits)
- `Name`s and `Tag`s are limited to 40 characters
- `Amount` is only specifiable up to 2 decimal places, and is capped at 1,000,000

### CODE REVIEW

File: src/main/java/seedu/address/model/person/Name.java [hunk 1]
```diff
@@ -10,13 +10,13 @@
...
     public static final String MESSAGE_CONSTRAINTS =
-            "Names should only contain alphanumeric characters and spaces, and it should not be blank";
+            "Names should only contain up to 40 alphanumeric characters and spaces, and it should not be blank";
 
     /*
      * The first character of the address must not be a whitespace,
      * otherwise " " (a blank string) becomes a valid input.
      */
-    public static final String VALIDATION_REGEX = "[\\p{Alnum}][\\p{Alnum} ]*";
+    public static final String VALIDATION_REGEX = "[\\p{Alnum}][\\p{Alnum} ]{1,39}";
```
Comment (Aulud) [inline src/main/java/seedu/address/model/person/Name.java]: This regex seems to exclude valid (although stupid) names with 1 character, e.g. `X`, i.e. the minimum length is 2. The previous implementation handled one-character names fine. Perhaps, the `{1, 39}` should be replaced with `{0, 39}` so the second capture block can capture up to 39 characters.

The validation does work for the Edit command on Contacts though, so that's good.

File: src/main/java/seedu/address/model/activity/Amount.java [hunk 2]
```diff
@@ -26,7 +27,7 @@ public Amount(double amount) {
      * Returns true if the given amount is a valid amount.
      */
     public static boolean isValidAmount(double test) {
-        return test > 0;
+        return test > 0 && test <= 1000000 && test * 100 == (int) (test * 100);
```
Comment (Aulud) [inline src/main/java/seedu/address/model/activity/Amount.java]: This looks like a pretty dodgy way to check if the valid amount is is specified to 2 d.p. - floating-point numbers may have different behaviour on different OSes (Windows for example, sometimes truncates trailing 0s from the decimal).

I think instead of limiting the decimal to exactly 2 d.p., we should just truncate/round the input to 2 d.p. (put this in the User Guide). Additionally, floating-point errors mean these two command strings behave very differently:
`expense e/25.50000000000001 p/Alex` - fails correctly
`expense e/25.500000000000001 p/Alex` - succeeds due to loss of precision
`expense e/999999.99999999999 p/Bernice` - succeeds due to loss of precision

File: src/main/java/seedu/address/model/person/Phone.java [hunk 3]
```diff
@@ -11,8 +11,8 @@
 
 
     public static final String MESSAGE_CONSTRAINTS =
-            "Phone numbers should only contain numbers, and it should be at least 3 digits long";
-    public static final String VALIDATION_REGEX = "\\d{3,}";
+            "Phone numbers should only contain numbers, and it should be between 3 to 20 digits long";
```
Comment (Aulud) [inline src/main/java/seedu/address/model/person/Phone.java]: Would `be 3 to 20 digits` be clearer since between brings up the debate on inclusive-of/exclusive-of?

File: src/test/java/seedu/address/model/person/NameTest.java [hunk 4]
```diff
@@ -29,12 +29,14 @@ public void isValidName() {
...
         assertFalse(Name.isValidName("peter*")); // contains non-alphanumeric characters
+        assertFalse(Name.isValidName("peterpeterpeterpeter peterpeterpeterpeter")); // contains 41 characters
 
         // valid name
         assertTrue(Name.isValidName("peter jack")); // alphabets only
         assertTrue(Name.isValidName("12345")); // numbers only
         assertTrue(Name.isValidName("peter the 2nd")); // alphanumeric characters
         assertTrue(Name.isValidName("Capital Tan")); // with capital letters
         assertTrue(Name.isValidName("David Roger Jackson Ray Jr 2nd")); // long names
+        assertTrue(Name.isValidName("peterpeterpeterpeterpeterpeterpeterpeter")); // contains 40 characters
```
Comment (Aulud) [inline src/test/java/seedu/address/model/person/NameTest.java:40]: LOL

**Review by Aulud — COMMENTED** [review]
Generally OK apart from a few corner cases - mind taking a look ASAP? >.< @liakify

File: src/main/java/seedu/address/model/person/Name.java (hunk 1 above)
Comment (liakify) [reply src/main/java/seedu/address/model/person/Name.java]: will change

File: src/main/java/seedu/address/model/person/Phone.java (hunk 3 above)
Comment (liakify) [reply src/main/java/seedu/address/model/person/Phone.java]: will do

File: src/main/java/seedu/address/model/activity/Amount.java (hunk 2 above)
Comment (liakify) [reply src/main/java/seedu/address/model/activity/Amount.java]: updated

**Review by daekoon — APPROVED** [review]
LGTM
